# 1. Reglas de asociación.

En el “groceries.csv” contiene las transacciones recogidas durante un mes en una tienda de comestibles. 
En cada una de las filas de este archivo se encuentran los artículos comprados de forma conjunta por los diferentes clientes. 

Obtener las lista de artículos que tiene un soporte mínimo de 0,15. 

Obtener también las reglas de asociación que se pueden deducir de este conjunto de datos con un soporte mínimo de 0,05 y confianza de 0,25.

### Importación de los datos:

Voy a importar los datos del fichero csv a una lista:

In [1]:
import csv

groceries=[]

groceries_file=csv.reader(open("groceries.csv", "rb"))

for row in groceries_file:

    groceries.append(row)


Compruebo cuantas observaciones están disponibles y veo las primeras.

In [2]:
print "Tenemos ", len(groceries), "observaciones. Las 5 primeras son:"
groceries[:5]# veo los 5 primeros registros

Tenemos  9835 observaciones. Las 5 primeras son:


[['citrus fruit', 'semi-finished bread', 'margarine', 'ready soups'],
 ['tropical fruit', 'yogurt', 'coffee'],
 ['whole milk'],
 ['pip fruit', 'yogurt', 'cream cheese ', 'meat spreads'],
 ['other vegetables',
  'whole milk',
  'condensed milk',
  'long life bakery product']]

Cada elemento de la lista es un producto o un conjunto de productos comprados en la tienda de comestibles.

### Lista de artículos que tiene un soporte mínimo de 0,15

Preparo una lista de observaciones con soporte mínimo de 0.15, es decir: los que aparecen como mínimo en 15% del total de las observaciones.

In [3]:
import apriori
F, soporte15 = apriori.apriori(groceries, min_support=0.15, verbose=True)

{soda}:  sup = 0.174
{whole milk}:  sup = 0.256
{other vegetables}:  sup = 0.193
{rolls/buns}:  sup = 0.184


Observamos que el producto más comprado es leche (whole milk - 25.6% de las compras), la verdura (19.3% de las compras), bollos y la sopa (18.5% y 17.4% de las compras).

### Reglas de asociación que se pueden deducir con un soporte mínimo de 0,05 y confianza de 0,25.

En primer lugar busco productos con el soporte mínimo de 5%.

In [21]:
F, soporte5 = apriori.apriori(groceries, min_support=0.05, verbose=True)

{citrus fruit}:  sup = 0.083
{curd}:  sup = 0.053
{root vegetables}:  sup = 0.109
{shopping bags}:  sup = 0.099
{bottled beer}:  sup = 0.081
{butter}:  sup = 0.055
{newspapers}:  sup = 0.08
{napkins}:  sup = 0.052
{pip fruit}:  sup = 0.076
{domestic eggs}:  sup = 0.063
{bottled water}:  sup = 0.111
{tropical fruit}:  sup = 0.105
{soda}:  sup = 0.174
{whole milk}:  sup = 0.256
{other vegetables}:  sup = 0.193
{canned beer}:  sup = 0.078
{frankfurter}:  sup = 0.059
{sausage}:  sup = 0.094
{yogurt}:  sup = 0.14
{pastry}:  sup = 0.089
{margarine}:  sup = 0.059
{coffee}:  sup = 0.058
{brown bread}:  sup = 0.065
{fruit/vegetable juice}:  sup = 0.072
{whipped/sour cream}:  sup = 0.072
{pork}:  sup = 0.058
{rolls/buns}:  sup = 0.184
{beef}:  sup = 0.052
{whole milk, yogurt}:  sup = 0.056
{whole milk, rolls/buns}:  sup = 0.057
{whole milk, other vegetables}:  sup = 0.075


Aparecen bastantes más productos que en la consulta anterior, porque el criterio es menos estricto.

Dentro de los productos con soporte mínimo eligo los conjuntos que cumplen con la confianza mínima 25%.

In [23]:
# Solo se mostrarán los items que tengan confianca 25% como mínimo.
H = apriori.generate_rules(F, soporte5, min_confidence=0.25, verbose=True)

{yogurt} ---> {whole milk}:  conf = 0.402, sup = 0.056
{rolls/buns} ---> {whole milk}:  conf = 0.308, sup = 0.057
{other vegetables} ---> {whole milk}:  conf = 0.387, sup = 0.075
{whole milk} ---> {other vegetables}:  conf = 0.293, sup = 0.075


Obtenemos cuatro conjuntos, de los que la asociación más significativa es la de yogurt y leche, aunque no es la que tiene mayor soporte. En 40.2% de los casos en los que se ha comprado ypgurt también se ha comprado leche y esta combinación de productos apareció en 5.6% de las compras.

Adicionalemnte observamos que:

En 38.7% de los casos en los que se ha comprado verdura también se ha comprado leche y esta combinación de productos apareció en 7.5% de las compras.
En 30.8% de los casos en los que se ha comprado bollos también se ha comprado leche y esta combinación de productos apareció en 5.7% de las compras.

Como conclusión general podemos decir que el producto más comprado - la leche - se suele comprar con yogurt, bollos y verduras 30%-40% de las veces que se compran esos tres productos.